### PR TITLE
Single-task POYO

### PR DESCRIPTION
**Pull request by vinamarora8**
Tested on entire MP dataset for up to 60 epochs. Works as expected. On 4xA40s: time per epoch: ~40s, time per validation: ~10s. I believe we are ready for training POYO-MP.

There is some ugliness wrt `subtask_index` handling. But I am deliberately leaving that until brainsets merges the interval-based approach so we don't have to re-engineer it later.

### CODE REVIEW

File: examples/poyo/configs/base.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+num_workers: 4
+seed: 42
+
+subtask_weights:
+  - 1.0  # RANDOM
+  - 0.1  # HOLD
+  - 5.0  # REACH
+  - 1.0  # RETURN
+  - 0.1  # INVALID
+  - 0.0  # OUTLIER
```
Comment (vinamarora8) [inline examples/poyo/configs/base.yaml:29]: This is hacky. Will clean once the interval-based movement phases are merged into brainsets.

**Comment by vinamarora8** [comment]
@mazabou Ready for review. If possible, we should merge this in before adding `examples/poyo/finetune.py`

**Review by mazabou — COMMENTED** [review]
Looks good overall. I left a few comments. 

I made the update to `brainsets`, I can attempt to incorporate them here.

File: examples/poyo/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# POYO 🧠
+Official codebase for POYO published at NeurIPS 2023
+[[project page]](https://poyo-brain.github.io/)
+[[arxiv]](https://arxiv.org/abs/2310.16046)
+
+### Training
+To train POYO-MP you can run:
+```bash
+python train.py --config-name train_poyo_mp.yaml
+```
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/README.md:10]: Yes, we should remember to do that

File: examples/poyo/configs/base.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+data_root: ./data/processed/
+log_dir: ./logs
+
+sequence_length: 1.0  # in seconds
+latent_step: 0.125  # in seconds
+modality_name: cursor_velocity_2d
+
+epochs: 1000  # if -1, will use steps
+steps: -1  # if -1, will use epochs. Epochs take precedence.
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/configs/base.yaml:11]: oh, I just copied these from poyo_plus I think. Although I added them there. Not needed for supervised cases. I'll remove it

[the remark this answers is not in the document]
Comment (mazabou) [reply examples/poyo/configs/base.yaml:11]: yes, it's not you who added it. I meant to ask this for a while. I know we had it before.

File: torch_brain/utils/stitcher.py [hunk 4]
```diff
@@ -71,7 +74,159 @@ def stitch(timestamps: torch.Tensor, values: torch.Tensor) -> torch.Tensor:
     return averages
 
 
-class StitchEvaluator(L.Callback):
+class MultiSessionDecodingStitchEvaluator(L.Callback):
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:77]: It is specific to multi-session situations because it requires the tokenizer to add `"session_id"` field to the batch-data. In single session situations users shouldn't be expected to add this field.

I believe having the `Stitcher` class (independent of Lightning, the idea we discussed on slack) would end up making this situation more unified. This would take some thinking and experimenting; which I can probably get to next week, but this week I'm tight on time. Until then, I feel this is a good solution.

[the remark this answers is not in the document]
Comment (mazabou) [reply torch_brain/utils/stitcher.py:77]: I agree with the second point. 

Re: `"session_id"`, I also agree. My solution is that instead of rewritting the whole class, we just add an argument to the existing one that would be "single_session=True". And it will not expect to get "session_id" in that case. everything else will be exactly the same.

File: torch_brain/utils/stitcher.py [hunk 5]
```diff
@@ -71,7 +74,159 @@ def stitch(timestamps: torch.Tensor, values: torch.Tensor) -> torch.Tensor:
...
+        self.on_validation_epoch_start(*args, **kwargs)
+
+    def on_test_batch_end(self, *args, **kwargs):
+        self.on_validation_batch_end(*args, **kwargs)
+
+    def on_test_epoch_end(self, *args, **kwargs):
+        self.on_validation_epoch_end(*args, **kwargs, prefix="test")
+
+
+class MultiSessionMultiTaskStitchEvaluator(L.Callback):
```
Comment (vinamarora8) [inline torch_brain/utils/stitcher.py:229]: Note to self: change the name to `MultiSessionMultiTaskDecodingStitchEvaluator`

File: torch_brain/utils/stitcher.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py]: Removed "MultiSession" from the name

File: torch_brain/models/poyo.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            output_weights = np.ones(len(output_values), dtype=np.float32)
+        else:
+            output_weights = self.subtask_weights[output_subtask_index]
+
+        # Mask for the output predictions
+        output_mask = np.ones(len(output_values), dtype=bool)
+        if self.eval:
+            # During eval, only evaluate on the subtask specified in the config
+            target_subtask_index = data.config["eval_subtask_index"]
+            output_mask = output_mask & (output_subtask_index == target_subtask_index)
```
Comment (mazabou) [inline torch_brain/models/poyo.py:329]: this will be replaced so I will not review it

File: torch_brain/models/poyo.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        subtask_weights (List[float]): Loss-weights for different subtasks.
+    """
+
+    def __init__(
+        self,
+        unit_tokenizer: Callable,
+        session_tokenizer: Callable,
+        latent_step: float,
+        num_latents_per_step: int,
+        modality_spec: ModalitySpec,
```
Comment (mazabou) [inline torch_brain/models/poyo.py:253]: should we name it `readout_spec`? `modality_spec` is more consistent, but could be confusing since spikes itself is a modality. and in the future we can change the input modality. so for consistency I would suggest calling this `readout_spec`

File: torch_brain/models/poyo.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+            # latent sequence
+            "latent_index": latent_index,
+            "latent_timestamps": latent_timestamps,
+            # output sequence
+            "output_session_index": pad8(session_index),
+            "output_timestamps": pad8(output_timestamps),
+            "output_mask": pad8(output_mask),
+            # ground truth targets
+            "target_values": pad8(output_values),
+            "target_weights": pad8(output_weights),
```
Comment (mazabou) [inline torch_brain/models/poyo.py:346]: here padding is used instead of chaining. how are you dealing with the loss when it comes to padding values?

**Review by mazabou — COMMENTED** [review]
Review of `torch_brain/models`

File: examples/poyo/train.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+from torch_brain.transforms import Compose
+
+# higher speed on machines with tensor cores
+torch.set_float32_matmul_precision("medium")
+
+logger = logging.getLogger(__name__)
+
+
+@hydra.main(version_base="1.3", config_path="./configs", config_name="train.yaml")
+def main(cfg: DictConfig):
```
Comment (mazabou) [inline examples/poyo/train.py:38]: the python convention is that `main()` goes at the bottom of the script. so this should be the last function.

File: examples/poyo/train.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+        target_weights = target_weights[output_mask]
+
+        if self.modality_spec.loss_fn == "mse":
+            loss = torch.nn.functional.mse_loss(
+                output_values, target_values, reduction="none"
+            )
+            loss = loss * target_weights[:, None]
+            loss = loss.mean()
+        else:
+            raise NotImplementedError("Only MSE loss is supported for now.")
```
Comment (mazabou) [inline examples/poyo/train.py:190]: why are we going backwards here?

File: torch_brain/utils/callbacks.py [hunk 11]
```diff
@@ -20,7 +20,7 @@ def on_train_epoch_start(self, trainer, pl_module):
     def on_train_epoch_end(self, trainer, pl_module):
         if self.enable:
             epoch_time = time.time() - self.start_time
-            pl_module.log("epoch_time", epoch_time)
+            pl_module.log("epoch_time", epoch_time, sync_dist=True)
```
Comment (mazabou) [inline torch_brain/utils/callbacks.py:23]: why do we need to sync_dist here? isn't everything blocked until the last gpu is done? also, why does this need to be accurate?

File: torch_brain/utils/stitcher.py [hunk 12]
```diff
@@ -71,7 +74,159 @@ def stitch(timestamps: torch.Tensor, values: torch.Tensor) -> torch.Tensor:
...
+    timestamps for each session. This cache is updated at the end of each batch.
+    Finally, at the end of the epoch, the cache is coalesced and the metric is
+    computed for each session.
+    """
+
+    def __init__(
+        self,
+        session_ids: Iterable[str],
+        modality_spec: Optional[ModalitySpec] = None,
+        metric_factory: Optional[Callable[[int], ModalitySpec]] = None,
```
Comment (mazabou) [inline torch_brain/utils/stitcher.py:118]: what's `metric_factory`? this nullifies the option to specify a different metric per dataset?

Comment (mazabou) [reply torch_brain/utils/stitcher.py:118]: are we creating a concept that will not work for POYOPlus?

File: examples/poyo/configs/train_poyo_mp.yaml [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+train_transforms:
+  - _target_: torch_brain.transforms.UnitDropout
+    max_units: 1000
+    min_units: 60
+    mode_units: 300
+    peak: 4
+
+wandb:
+  run_name: poyo_mp_mini
```
Comment (mazabou) [inline examples/poyo/configs/train_poyo_mp.yaml:15]: why mini?

**Review by mazabou — COMMENTED** [review]
Review part 2 of 2

File: torch_brain/models/poyo.py (hunk 7 above)
Comment (vinamarora8) [reply torch_brain/models/poyo.py:253]: We should change this in the spec python file as well in that case. I just went with whatever name was used there. I support renaming it to readout_spec at both places.

File: examples/poyo/train.py (hunk 9 above)
Comment (vinamarora8) [reply examples/poyo/train.py:38]: yeah we can rename this to `train()` or smth. Otherwise I don't like having both DataModule and TrainWrapper at the top. Too much code before you get to what this script is actually running. Happy to reconsider.

File: examples/poyo/train.py (hunk 10 above)
Comment (vinamarora8) [reply examples/poyo/train.py:190]: Just wanted to get it done without having to test _this_ code on datasets with other kinds of losses. We can bring back other losses when we've tested them. Imo we shouldn't include a feature we haven't tested. Adding more losses can be separate PRs.

File: torch_brain/utils/callbacks.py (hunk 11 above)
Comment (vinamarora8) [reply torch_brain/utils/callbacks.py:23]: It's not needed in theory. Lightning complains if we don't have sync_dist=True. I'd rather not see that warning. This incurs negligible cost.

File: torch_brain/utils/stitcher.py (hunk 12 above)
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:118]: `metric_factory` is when you wanna use a custom metric bypassing the modality_spec framework.

Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:118]: This will not work for POYOPlus. This stitcher is not meant for handling multi-task output

Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:118]: I think I have made a mistake here. metric_factor's type should be `Optional[Callable[[int], torchmetrics.Metric]`

File: examples/poyo/configs/train_poyo_mp.yaml (hunk 13 above)
Comment (vinamarora8) [reply examples/poyo/configs/train_poyo_mp.yaml:15]: Idk, ill call it poyo_mp

File: torch_brain/models/poyo.py [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+    recordings.
+
+    TODO: Document the model architecture
+
+    Args:
+        dim: Hidden dimension of the model
+        dim_out: Dimension of the decoding targets
+        dim_head: Dimension of each attention head
+        num_latents: Number of unique latent tokens (repeated at every latent step)
+        depth: Number of processing layers (self-attentions in the latent space)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/models/poyo.py:39]: Sure, it can be a separate PR focused on POYOPlus edits.

File: torch_brain/models/poyo.py [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+    <https://arxiv.org/abs/2310.16046>`.
+
+    POYO is a transformer-based model for neural decoding from electrophysiological
+    recordings.
+
+    TODO: Document the model architecture
+
+    Args:
+        dim: Hidden dimension of the model
+        dim_out: Dimension of the decoding targets
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/models/poyo.py]: done

File: torch_brain/models/__init__.py [hunk 16]
```diff
@@ -1 +1,2 @@
 from .poyo_plus import POYOPlus, POYOPlusTokenizer
+from .poyo import POYO, poyo_mp, POYOTokenizer
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/models/__init__.py:1]: done

**Comment by vinamarora8** [comment]
Note I have removed `strategy=ddp_find_unused_parameters_true ...` and `accelerator = ...` arguments of lightning trainer because we don't need them for this code.

File: torch_brain/models/poyo.py [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+        input_token_type: TensorType["batch", "n_in", int],
+        input_mask: Optional[TensorType["batch", "n_in", bool]] = None,
+        # latent sequence
+        latent_index: TensorType["batch", "n_latent", int],
+        latent_timestamps: TensorType["batch", "n_latent", float],
+        # output sequence
+        output_session_index: TensorType["batch", "n_out", int],
+        output_timestamps: TensorType["batch", "n_out", float],
+    ) -> TensorType["batch", "n_out", "dim_out", float]:
+        """Forward pass of the POYO+ model.
```
[the remark this answers is not in the document]
Comment (mazabou) [reply torch_brain/models/poyo.py]: To clarify, I am not referring to chain vs. pad. What I am suggesting is replicating the `unpack_output=True` behavior, so that instead of returning `B x T_out x D_out`, we return `List[*, D_out]` of length `B`. The logic for this is super simple. The typical user is not aware that masking is being used! — it's happening in the tokenizer/collate. so they wouldn't know that the output they are getting will need to be "unpadded", in which case they would have to know how to access to the `output_mask` inside of the batch dict. This would of course happen rarely but it is likely incase some dataset has missing values etc...

My suggestion is to include a single line of code:
```
if unpack_output:
     return [output[b][output_mask[b]] for b in range(output.size(0))]
```

File: torch_brain/models/poyo.py (hunk 7 above)
Comment (mazabou) [reply torch_brain/models/poyo.py:253]: what do you mean by the spec python file? you mean renaming `ModalitySpec` to `ReadoutSpec`?

File: examples/poyo/train.py (hunk 9 above)
Comment (mazabou) [reply examples/poyo/train.py:38]: let's put it back the way it was please. if you want to make the change across the board, it should be a separate PR/ a separate discussion.

File: examples/poyo/train.py (hunk 10 above)
Comment (mazabou) [reply examples/poyo/train.py:186]: we have `compute_loss_or_metric` that we can use no?

File: torch_brain/utils/stitcher.py (hunk 12 above)
Comment (mazabou) [reply torch_brain/utils/stitcher.py]: this is a new concept. I would rather it be in a separate PR/discussion. For now, I do not see why are introducing two completely different ways for using the stitcher, based on whether it's POYO or POYOPlus. It will be extremely confusing.

Comment (vinamarora8) [reply torch_brain/utils/stitcher.py]: is your issue with the existence of `metric_factory` as an optional parameter?

File: examples/poyo/train.py (hunk 10 above)
Comment (vinamarora8) [reply examples/poyo/train.py]: forgot about compute_loss_or_metric. fixed.

File: torch_brain/models/poyo.py (hunk 7 above)
Comment (vinamarora8) [reply torch_brain/models/poyo.py:253]: Yes

File: torch_brain/models/poyo.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/models/poyo.py]: Fair point. Done.

File: torch_brain/utils/stitcher.py (hunk 12 above)
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py]: removed

File: torch_brain/models/poyo.py (hunk 7 above)
Comment (vinamarora8) [reply torch_brain/models/poyo.py]: Renamed `modality_spec` to `readout_spec`. Done.

**Review by mazabou — APPROVED** [review]
Look great! Thank you for the colossal amount of work. Good to merge.

Just checking, you are going to merge this into main correct?
